grants_tagger_light/retagging/retagging.py

```diff
@@ -0,0 +1,319 @@
+            with open(curation_file, "w") as f:
+                json.dump(human_supervision, f)
+        count = len(human_supervision[tag][field])
+
+
+def _curate(save_to_path, pos_dset, neg_dset, tag, limit):
+    logging.info("- Curating positive examples")
+    _annotate(save_to_path, pos_dset, tag, limit, is_positive=True)
+
+    logging.info("- Curating negative examples")
+    _annotate(save_to_path, neg_dset, tag, limit, is_positive=False)
+
+
+def retag(
+    data_path: str,
+    save_to_path: str,
+    num_proc: int = os.cpu_count(),
+    batch_size: int = 1024,
+    tags: list = None,
+    tags_file_path: str = None,
+    threshold: float = 0.8,
+    train_examples: int = 100,
+    supervised: bool = True,
+    years: list = None,
+):
+    # We only have 1 file, so no sharding is available https://huggingface.co/docs/datasets/loading#multiprocessing
+    logging.info("Loading the MeSH jsonl...")
+    dset = load_dataset("json", data_files=data_path, num_proc=1)
+    if "train" in dset:
+        dset = dset["train"]
+
+    if years is not None:
+        logger.info(f"Removing all years which are not in {years}")
+        dset = dset.filter(
+            lambda x: any(np.isin(years, [str(x["year"])])), num_proc=num_proc
+        )
+
+    if tags_file_path is not None and os.path.isfile(tags_file_path):
+        with open(tags_file_path, "r") as f:
+            tags = [x.strip() for x in f.readlines()]
+
+    logging.info(f"- Total tags detected: {tags}.")
+    logging.info("- Training classifiers (retaggers)")
+
+    for tag in tags:
+        os.makedirs(os.path.join(save_to_path, tag.replace(" ", "")), exist_ok=True)
+        logging.info(f"- Obtaining positive examples for {tag}...")
+        positive_dset = dset.filter(lambda x: tag in x["meshMajor"], num_proc=num_proc)
+
+        if len(positive_dset["abstractText"]) < train_examples:
+            logging.info(
+                f"Skipping {tag}: low examples "
+                f"({len(positive_dset['abstractText'])} vs "
+                f"expected {train_examples}). "
+                f"Check {save_to_path}.err for more information "
+                f"about skipped tags."
+            )
+            with open(f"{save_to_path}.err", "a") as f:
+                f.write(tag)
+            continue
+        logging.info(f"-- Total positive examples for {tag}: {len(positive_dset)}")
+        logging.info(f"- Obtaining negative examples ('other') for {tag}...")
+        negative_dset = dset.filter(
+            lambda x: tag not in x["meshMajor"], num_proc=num_proc
+        )
+        logging.info(f"-- Total negative examples for {tag}: {len(negative_dset)}")
+
+        curation_file = os.path.join(save_to_path, tag.replace(" ", ""), "curation")
+        if supervised:
+            logging.info(f"- Curating {tag}...")
+            _curate(curation_file, positive_dset, negative_dset, tag, train_examples)
+        else:
```

Comment:
what the point of the `else` if you dont curate you train using the existing data

Reply:
Getting a random sample of for example 100 already provides some improvement, even not curating. I did a check for Artificial Intelligence. Not curating - F1 of 0.85. Curation - F1 of 0.95. I think it may be worth to leave it.